module/zfs/dsl_dataset.c

```diff
@@ -3087,7 +3087,11 @@ dsl_dataset_handoff_check(dsl_dataset_t *ds, void *owner, dmu_tx_t *tx)
 		dsl_dataset_long_rele(ds, owner);
 	}
 
-	held = dsl_dataset_long_held(ds);
+	dsl_dir_t *dd = ds->ds_dir;
+	mutex_enter(&dd->dd_activity_lock);
+	if (zfs_refcount_count(&ds->ds_longholds) != dd->dd_activity_count)
```

Comment:
since we're now getting the hold count, could we avoid the long_rele/long_hold, instead subtracting one from ds_longholds if owner is set?